ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -706,204 +388,100 @@ UIGestureRecognizerDelegate {
 
   /// - Returns: The index of the keyboard if it exist in user keyboards list
   public func indexForUserKeyboard(withID keyboardID: String?, languageID: String?) -> Int? {
-    let userKeyboards = activeUserDefaults().array(forKey: Key.userKeyboardsList)
-    guard let keyboards = userKeyboards as? [[String: String]] else {
-      return nil
-    }
-    return keyboards.index { kb in keyboardID == kb[Key.keyboardId] && languageID == kb[Key.languageId] }
+    let userKeyboards = activeUserDefaults().installableKeyboards(forKey: Key.userKeyboardsList)
+    return userKeyboards?.index { $0.id == keyboardID && $0.languageID == languageID }
+  }
+
+  public func userKeyboard(withID keyboardID: String, languageID: String) -> InstallableKeyboard? {
+    let userKeyboards = activeUserDefaults().installableKeyboards(forKey: Key.userKeyboardsList)
+    return userKeyboards?.first { $0.id == keyboardID && $0.languageID == languageID }
```

Comment:
Or maybe these particular functions here should be moved to the UserDefaults section?

Reply:
I didn't put it in `UserDefaults` because frequently functions perform several operations on the list of keyboards. So if these were in `UserDefaults`, we'd be reading and deserializing from `UserDefaults` multiple times, which is not ideal. We could extend `[InstallableKeyboard]` with some helpers though.

I was also thinking of having a struct containing keyboard and language ID so that `keyboardsDictionary` could be keyed by this instead of just a String. If so then this would be further simplified and then maybe a helper isn't necessary.

Reply:
I think you misunderstand.  Why not literally lift this function and place it within the `UserDefaults` class as an optional helper function?  Or would the same effect be gained by extending `[InstallableKeyboard]`?

Reply:
I think you misunderstand. Why not literally lift this function and place it within the `UserDefaults` class as an optional helper function? Or would the same effect be gained by extending `[InstallableKeyboard]`?

Reply:
I agree that this function is better placed in `UserDefaults` than in `Manager`. My point was that it's more flexible to extend `[InstallableKeyboard]` with the helpers than to extend `UserDefaults`. Although the flexibility is likely unneeded, now that I think about it, because the main usage would be to read a single keyboard, and the operation is pretty straightforward anyway.

On the same note, `userKeyboards` and `currentKeyboard` should probably be properties of `UserDefaults` as well.